File: src/backend/zuno/api/dto/workspace.py
from enum import Enum
from typing import Any, List, Literal, TypedDict

from pydantic import BaseModel, Field


WorkspaceProductMode = Literal[
    "enterprise_kb",
    "hr_resume",
    "contract_review",
    "general_agent",
]
WorkspaceRetrievalProfile = Literal["standard", "deep"]
WorkspaceTaskStatus = Literal[
    "created",
    "context_building",
    "planning",
    "running",
    "approval_waiting",
    "approval_required",
    "resuming",
    "finalizing",
    "recoverable_failed",
    "completed",
    "failed",
    "cancelled",
]
WorkspaceTaskLifecycleState = Literal[
    "pending",
    "running",
    "approval_required",
    "recoverable_failed",
    "cancelled",
    "completed",
]
WORKSPACE_TASK_LIFECYCLE_FLOW = [
    "pending",
    "running",
    "approval_required",
    "recoverable_failed",
    "cancelled",
    "completed",
]
WORKSPACE_TASK_STATUS_FLOW = [
    "created",
    "context_building",
    "planning",
    "running",
    "approval_waiting",
    "approval_required",
    "resuming",
    "finalizing",
    "recoverable_failed",
    "completed",
    "failed",
    "cancelled",
]
WORKSPACE_TASK_STATUS_TO_LIFECYCLE = {
    "created": "pending",
    "context_building": "pending",
    "planning": "running",
    "running": "running",
    "approval_waiting": "approval_required",
    "approval_required": "approval_required",
    "resuming": "running",
    "finalizing": "running",
    "recoverable_failed": "recoverable_failed",
    "failed": "recoverable_failed",
    "cancelled": "cancelled",
    "completed": "completed",
}
WORKSPACE_TASK_RECOVERY_ACTIONS = {
    "recoverable_failed": ["retry_task", "download_trace", "send_feedback"],
    "cancelled": ["retry_task", "send_feedback"],
    "completed": ["download_artifact", "send_feedback"],
}


class WorkSpaceAgents(Enum):
    SimpleAgent: str = "simple"
    WeChatAgent: str = "wechat-agent"


class WorkspaceTaskBudget(BaseModel):
    max_steps: int | None = None
    max_tokens: int | None = None
    timeout_seconds: int | None = None
    cost_ceiling: float | None = None


class WorkspaceOutputContract(BaseModel):
    artifact_kinds: list[str] = Field(default_factory=list)
    citation_required: bool = True
    trace_required: bool = True
    format: str = "markdown"


class KnowledgeSpaceRetrievalSelection(BaseModel):
    knowledge_space_id: str
    retrieval_profile: WorkspaceRetrievalProfile = "standard"


class WorkSpaceSimpleTask(BaseModel):
    query: str
    model_id: str
    session_id: str
    workspace_id: str | None = None
    user_id: str | None = None
    task_id: str | None = None
    trace_id: str | None = None
    goal: str | None = None
    product_mode: WorkspaceProductMode = "general_agent"
    knowledge_space_ids: List[str] = Field(default_factory=list)
    knowledge_space_profiles: List[KnowledgeSpaceRetrievalSelection] = Field(default_factory=list)
    retrieval_profiles: dict[str, WorkspaceRetrievalProfile] = Field(default_factory=dict)
    uploaded_file_ids: List[str] = Field(default_factory=list)
    approval_mode: str = "auto"
    budget: WorkspaceTaskBudget | None = None
    output_contract: WorkspaceOutputContract | None = None
    workspace_mode: str = "normal"
    agent_name: str | None = None
    agent_id: str | None = None
    web_search: bool = True
    plugins: List[str] = Field(default_factory=list)
    mcp_servers: List[str] = Field(default_factory=list)
    knowledge_ids: List[str] = Field(default_factory=list)
    retrieval_mode: str = "auto"
    multi_agent_enabled: bool = False
    agent_skill_ids: List[str] = Field(default_factory=list)
    execution_mode: str = "tool"
    access_scope: str = "workspace"
    desktop_bridge_url: str | None = None
    desktop_bridge_token: str | None = None
    attachments: List["WorkspaceAttachment"] = Field(default_factory=list)


class WorkspaceAttachment(BaseModel):
    name: str
    url: str
    mime_type: str | None = None
    size: int | None = None


class WorkspaceProductObjectBase(BaseModel):
    workspace_id: str
    owner: str | None = None
    status: str = "active"
    policy_scope: str = "workspace"
    trace_id: str | None = None
    created_at: str | None = None
    updated_at: str | None = None
    retention_policy: str | None = None


class WorkspaceContract(WorkspaceProductObjectBase):
    members: list[str] = Field(default_factory=list)


class KnowledgeSpaceContract(WorkspaceProductObjectBase):
    knowledge_space_id: str
    graph_project_id: str | None = None
    index_version: str | None = None
    acl_policy: str = "workspace"


class KnowledgeSpaceConfig(BaseModel):
    name: str
    description: str = ""
    workspace_id: str
    acl_scope: str = "workspace"
    default_sensitivity: str = "internal"
    index_capabilities: dict[str, Any] = Field(default_factory=dict)
    parser_config: dict[str, Any] = Field(default_factory=dict)
    chunk_config: dict[str, Any] = Field(default_factory=dict)
    embedding_config: dict[str, Any] = Field(default_factory=dict)
    graph_config: dict[str, Any] = Field(default_factory=dict)
    ocr_vlm_config: dict[str, Any] = Field(default_factory=dict)
    retrieval_defaults: dict[str, Any] = Field(
        default_factory=lambda: {
            "default_profile": "standard",
            "available_profiles": ["standard", "deep"],
        }
    )
    security_policy: dict[str, Any] = Field(default_factory=dict)


class WorkspaceSessionContract(WorkspaceProductObjectBase):
    session_id: str
    user_id: str
    thread_id: str | None = None
    active_task_id: str | None = None


class WorkspaceTaskContract(WorkspaceProductObjectBase):
    task_id: str
    session_id: str
    goal: str
    product_mode: WorkspaceProductMode = "general_agent"
    status: WorkspaceTaskStatus = "created"
    budget: WorkspaceTaskBudget = Field(default_factory=WorkspaceTaskBudget)


class WorkspaceTaskLifecycleSnapshot(BaseModel):
    task_id: str
    trace_id: str | None = None
    state: WorkspaceTaskLifecycleState
    status: str
    recoverable: bool = False
    recovery_actions: list[str] = Field(default_factory=list)
    downloadable_artifact_ids: list[str] = Field(default_factory=list)


class UploadedFileContract(WorkspaceProductObjectBase):
    file_id: str
    mime_type: str
    hash: str
    security_label: str = "internal"
    parse_status: str = "created"


class WorkspaceFileStatus(BaseModel):
    file_id: str
    filename: str | None = None
    mime_type: str
    size_bytes: int = 0
    source_sha256: str
    storage_uri: str | None = None
    source_ref: str | None = None
    parse_status: str
    index_status: str = "not_indexed"
    parser_id: str | None = None
    document_version_id: str | None = None
    index_job_id: str | None = None
    blocked_reason: str | None = None
    dependency_probe: dict[str, Any] = Field(default_factory=dict)
    retry_count: int = 0
    last_error: str | None = None
    actions: list[str] = Field(
        default_factory=lambda: [
            "retry",
            "cancel",
            "reparse",
            "reindex",
            "rebuild_graph",
            "view_diagnostics",
        ]
    )


class WorkspaceCitationRef(BaseModel):
    citation_id: str
    evidence_id: str
    document_id: str
    block_id: str
    source_ref: str
    source_span: dict[str, Any] = Field(default_factory=dict)
    trust_label: str | None = None
    source_uri: str | None = None
    provenance: dict[str, Any] = Field(default_factory=dict)


class ArtifactContract(WorkspaceProductObjectBase):
    artifact_id: str
    task_id: str
    kind: str
    uri: str
    hash: str | None = None
    download_policy: str = "workspace"
    citation_refs: list[WorkspaceCitationRef] = Field(default_factory=list)


class ChangeImpactPreview(BaseModel):
    change_type: str
    triggered_action: str
    affected_file_count: int = 0
    affected_chunk_count: int = 0
    affects_existing_artifacts: bool = False
    requires_external_provider: bool = False
    may_create_blocked_state: bool = False
    estimated_duration_ms: int | None = None
    user_visible_summary: str | None = None


class TraceEventContract(BaseModel):
    event_id: str
    task_id: str
    trace_id: str
    type: str
    timestamp: float
    payload: dict[str, Any] = Field(default_factory=dict)


class CitationContract(BaseModel):
    citation_id: str
    evidence_id: str
    document_id: str
    block_id: str
    source_span: dict[str, Any] = Field(default_factory=dict)
    created_at: str | None = None


class FeedbackContract(BaseModel):
    feedback_id: str
    task_id: str
    rating: int | None = None
    label: str | None = None
    comment: str | None = None
    dataset_candidate: bool = False
    created_at: str | None = None


class WorkspaceProductStreamEvent(BaseModel):
    event_id: str
    task_id: str
    trace_id: str
    type: str
    status: WorkspaceTaskStatus | str
    timestamp: float
    payload: dict[str, Any] = Field(default_factory=dict)


class WorkspaceAgentEventData(TypedDict, total=False):
    event_id: str
    task_id: str
    trace_id: str
    artifact_id: str
    citation_ids: list[str]
    phase: str
    status: str
    message: str
    tool_name: str
    tool_type: str
    tool_call_id: str
    arguments: dict[str, Any]
    result: str
    chunk: str
    accumulated: str
    error: str
    done: bool
    retrieval_mode: str
    first_mode: str
    final_mode: str
    fallback_reason: str
    round_count: int
    second_pass_used: bool
    rewritten_query_used: bool
    query_variants: list[str]
    rounds: list[dict[str, Any]]


class WorkspaceAgentStreamEvent(TypedDict):
    event: Literal["status", "tool_call", "tool_result", "final"]
    timestamp: float
    data: WorkspaceAgentEventData


__all__ = [
    "ArtifactContract",
    "ChangeImpactPreview",
    "CitationContract",
    "FeedbackContract",
    "KnowledgeSpaceConfig",
    "KnowledgeSpaceContract",
    "KnowledgeSpaceRetrievalSelection",
    "TraceEventContract",
    "UploadedFileContract",
    "WorkspaceCitationRef",
    "WorkspaceFileStatus",
    "WORKSPACE_TASK_STATUS_FLOW",
    "WorkSpaceAgents",
    "WorkSpaceSimpleTask",
    "WorkspaceAgentEventData",
    "WorkspaceAgentStreamEvent",
    "WorkspaceAttachment",
    "WorkspaceContract",
    "WorkspaceOutputContract",
    "WorkspaceProductObjectBase",
    "WorkspaceProductStreamEvent",
    "WorkspaceRetrievalProfile",
    "WorkspaceSessionContract",
    "WorkspaceTaskBudget",
    "WorkspaceTaskContract",
    "WorkspaceTaskLifecycleSnapshot",
    "WorkspaceTaskLifecycleState",
    "WORKSPACE_TASK_LIFECYCLE_FLOW",
    "WORKSPACE_TASK_RECOVERY_ACTIONS",
    "WORKSPACE_TASK_STATUS_TO_LIFECYCLE",
]

File: apps/web/src/apis/workspace.ts
import { fetchEventSource } from '@microsoft/fetch-event-source'
import { request } from '../utils/request'
import { apiUrl } from '../utils/api'

export interface ExecutionModeDefinition {
  id: string
  label: string
  summary: string
  capabilities: string[]
  restrictions: string[]
  supports_tools: boolean
  supports_terminal: boolean
}

export interface AccessScopeDefinition {
  id: string
  label: string
  summary: string
  capabilities: string[]
  restrictions: string[]
  risk_level: string
}

export interface WorkspaceExecutionConfig {
  execution_modes: ExecutionModeDefinition[]
  access_scopes: AccessScopeDefinition[]
}

export const getWorkspaceExecutionModesAPI = async () => {
  return request({
    url: '/api/v1/workspace/execution-modes',
    method: 'get',
  })
}

export const getWorkspacePluginsByModeAPI = async (executionMode: string, accessScope: string) => {
  return request({
    url: '/api/v1/workspace/plugins',
    method: 'get',
    params: {
      execution_mode: executionMode,
      access_scope: accessScope,
    },
  })
}

export const getWorkspaceSessionsAPI = async (workspaceMode?: string) => {
  return request({
    url: '/api/v1/workspace/session',
    method: 'get',
    params: workspaceMode
      ? {
          workspace_mode: workspaceMode,
        }
      : undefined,
  })
}

export const createWorkspaceSessionAPI = async (data: {
  title?: string
  contexts?: any[]
  session_id?: string
  agent?: string
  workspace_mode?: 'normal' | 'agent'
}) => {
  return request({
    url: '/api/v1/workspace/session',
    method: 'post',
    data,
  })
}

export const getWorkspaceSessionInfoAPI = async (sessionId: string) => {
  return request({
    url: `/api/v1/workspace/session/${sessionId}`,
    method: 'post',
  })
}

export const deleteWorkspaceSessionAPI = async (sessionId: string) => {
  return request({
    url: '/api/v1/workspace/session',
    method: 'delete',
    params: {
      session_id: sessionId,
    },
  })
}

export type WorkspaceProductMode = 'enterprise_kb' | 'hr_resume' | 'contract_review' | 'general_agent'

export type WorkspaceTaskStatus =
  | 'created'
  | 'pending'
  | 'context_building'
  | 'planning'
  | 'running'
  | 'approval_waiting'
  | 'approval_required'
  | 'resuming'
  | 'finalizing'
  | 'recoverable_failed'
  | 'completed'
  | 'failed'
  | 'cancelled'

export type WorkspaceTaskLifecycleState =
  | 'pending'
  | 'running'
  | 'approval_required'
  | 'recoverable_failed'
  | 'cancelled'
  | 'completed'

export interface WorkspaceTaskBudget {
  max_steps?: number
  max_tokens?: number
  timeout_seconds?: number
  cost_ceiling?: number
}

export interface WorkspaceOutputContract {
  artifact_kinds: string[]
  citation_required: boolean
  trace_required: boolean
  format?: string
}

export type WorkspaceRetrievalProfile = 'standard' | 'deep'

export interface KnowledgeSpaceRetrievalSelection {
  knowledge_space_id: string
  retrieval_profile: WorkspaceRetrievalProfile
}

export interface WorkspaceProductObjectBase {
  workspace_id: string
  owner?: string
  status: string
  policy_scope?: string
  trace_id?: string
  created_at?: string
  updated_at?: string
  retention_policy?: string
}

export interface WorkspaceContract extends WorkspaceProductObjectBase {
  members: string[]
}

export interface KnowledgeSpaceContract extends WorkspaceProductObjectBase {
  knowledge_space_id: string
  graph_project_id?: string
  index_version?: string
  acl_policy?: string
}

export interface KnowledgeSpaceConfig {
  name: string
  description?: string
  workspace_id: string
  acl_scope?: string
  default_sensitivity?: string
  index_capabilities?: Record<string, any>
  parser_config?: Record<string, any>
  chunk_config?: Record<string, any>
  embedding_config?: Record<string, any>
  graph_config?: Record<string, any>
  ocr_vlm_config?: Record<string, any>
  retrieval_defaults?: {
    default_profile?: WorkspaceRetrievalProfile
    available_profiles?: WorkspaceRetrievalProfile[]
    [key: string]: any
  }
  security_policy?: Record<string, any>
}

export interface WorkspaceSessionContract extends WorkspaceProductObjectBase {
  session_id: string
  user_id: string
  thread_id?: string
  active_task_id?: string
}

export interface WorkspaceTaskContract extends WorkspaceProductObjectBase {
  task_id: string
  session_id: string
  goal: string
  product_mode: WorkspaceProductMode
  status: WorkspaceTaskStatus
  budget?: WorkspaceTaskBudget
}

export interface UploadedFileContract extends WorkspaceProductObjectBase {
  file_id: string
  mime_type: string
  hash: string
  security_label?: string
  parse_status: string
}

export interface WorkspaceFileStatus {
  file_id: string
  filename?: string
  mime_type: string
  size_bytes: number
  source_sha256: string
  storage_uri?: string
  source_ref?: string
  parse_status: string
  index_status: string
  parser_id?: string
  document_version_id?: string
  index_job_id?: string
  blocked_reason?: string
  dependency_probe: Record<string, any>
  retry_count: number
  last_error?: string
  actions: string[]
}

export interface WorkspaceCitationRef {
  citation_id: string
  evidence_id: string
  document_id: string
  block_id: string
  source_ref: string
  source_span: Record<string, any>
  trust_label?: string
  source_uri?: string
  provenance?: Record<string, any>
}

export interface ArtifactContract extends WorkspaceProductObjectBase {
  artifact_id: string
  task_id: string
  kind: string
  uri: string
  hash?: string
  download_policy?: string
  citation_refs: WorkspaceCitationRef[]
}

export interface ChangeImpactPreview {
  change_type: string
  triggered_action: string
  affected_file_count: number
  affected_chunk_count: number
  affects_existing_artifacts: boolean
  requires_external_provider: boolean
  may_create_blocked_state: boolean
  estimated_duration_ms?: number
  user_visible_summary?: string
}

export interface WorkspaceTaskLifecycleSnapshot {
  task_id: string
  trace_id?: string
  state: WorkspaceTaskLifecycleState
  status: string
  recoverable: boolean
  recovery_actions: string[]
  downloadable_artifact_ids: string[]
}

export interface TraceEventContract {
  event_id: string
  task_id: string
  trace_id: string
  type: string
  timestamp: number
  payload: Record<string, any>
}

export interface CitationContract {
  citation_id: string
  evidence_id: string
  document_id: string
  block_id: string
  source_span: Record<string, any>
  created_at?: string
}

export interface FeedbackContract {
  feedback_id: string
  task_id: string
  rating?: number
  label?: string
  comment?: string
  dataset_candidate?: boolean
  created_at?: string
}

export interface WorkSpaceSimpleTask {
  query: string
  model_id: string
  workspace_id?: string
  user_id?: string
  task_id?: string
  trace_id?: string
  goal?: string
  product_mode?: WorkspaceProductMode
  knowledge_space_ids?: string[]
  knowledge_space_profiles?: KnowledgeSpaceRetrievalSelection[]
  retrieval_profiles?: Record<string, WorkspaceRetrievalProfile>
  uploaded_file_ids?: string[]
  approval_mode?: string
  budget?: WorkspaceTaskBudget
  output_contract?: WorkspaceOutputContract
  workspace_mode?: string
  agent_name?: string
  agent_id?: string
  web_search?: boolean
  plugins: string[]
  mcp_servers: string[]
  knowledge_ids?: string[]
  retrieval_mode?: string
  agent_skill_ids?: string[]
  session_id?: string
  execution_mode: string
  access_scope: string
  desktop_bridge_url?: string
  desktop_bridge_token?: string
  attachments?: WorkspaceAttachment[]
}

export interface WorkspaceTaskCreateResponse {
  task: WorkspaceTaskContract
  artifact_ids: string[]
  artifacts: ArtifactContract[]
  feedback_ids?: string[]
  feedback?: FeedbackContract[]
  lifecycle?: WorkspaceTaskLifecycleSnapshot
  runtime?: WorkspaceRuntimeSnapshot
  observability?: WorkspaceObservabilitySnapshot
  retrieval_plan?: Record<string, any>
  plan_summary?: Record<string, any>
  reflection_summary?: Record<string, any>
  replan_summary?: Record<string, any>
  trace_summary?: Record<string, any>
  eval_summary?: Record<string, any>
  cost_summary?: Record<string, any>
  capability_snapshot?: Record<string, any>
  knowledge_config_summary?: Record<string, any>
}

export interface WorkspaceArtifactResponse {
  artifact: ArtifactContract
  content: string
  citation_refs: WorkspaceCitationRef[]
  download?: {
    url: string
    filename: string
    media_type: string
    policy: string
  }
}

export interface WorkspaceFeedbackRequest {
  task_id: string
  rating?: number
  label?: string
  comment?: string
  dataset_candidate?: boolean
}

export interface WorkspaceFileCreateRequest {
  workspace_id: string
  file_id?: string
  name?: string
  mime_type: string
  hash?: string
  uri?: string
  content?: string
  trace_id?: string
  security_label?: string
}

export interface WorkspaceFileCreateResponse {
  file: UploadedFileContract
  name?: string
  uri?: string
  file_status?: WorkspaceFileStatus
}

export interface WorkspaceIngestRequest {
  workspace_id: string
  file_id: string
  knowledge_space_id: string
  session_id?: string
  trace_id?: string
}

export interface WorkspaceIngestResponse {
  ingest_task_id: string
  workspace_id: string
  file_id: string
  knowledge_space_id: string
  session_id?: string
  trace_id: string
  status: string
  file: UploadedFileContract
  file_status?: WorkspaceFileStatus
}

export interface WorkspaceApprovalRequest {
  decision: 'approved' | 'rejected'
  comment?: string
  approval_id?: string
  tool_call_id?: string
  required_approval?: string
}

export interface WorkspaceApprovalResponse extends WorkspaceTaskCreateResponse {}

export interface WorkspaceCancelRequest {
  reason?: string
}

export interface WorkspaceTaskLifecycleResponse {
  states: WorkspaceTaskLifecycleState[]
  terminal_states: WorkspaceTaskLifecycleState[]
  status_mapping: Record<string, WorkspaceTaskLifecycleState>
  recovery_actions: Record<string, string[]>
}

export interface WorkspaceRuntimeSnapshot {
  task_id: string
  trace_id: string
  thread_id: string
  workspace_id: string
  status: WorkspaceTaskStatus | string
  state: Record<string, any>
  checkpoint_ids: string[]
  latest_checkpoint?: Record<string, any> | null
  pending_interrupt?: Record<string, any> | null
  failure?: Record<string, any> | null
  events: Array<Record<string, any>>
}

export interface WorkspaceObservabilitySnapshot {
  spans: Array<Record<string, any>>
  release_eval?: Record<string, any> | null
  trace_replay: {
    source_refs: string[]
    event_ids?: string[]
    trace_id?: string
    task_id?: string
  }
}

export interface WorkspaceAttachment {
  name: string
  url: string
  mime_type?: string
  size?: number
}

export interface WorkspaceStreamEvent {
  id: string
  type: string
  title: string
  detail: string
  isFinal?: boolean
  task_id?: string
  trace_id?: string
  artifact_id?: string
  citation_ids?: string[]
  tool_id?: string
  tool_call_id?: string
  approval_id?: string
  required_approval?: string
  audit_ref?: string
  lifecycle_state?: WorkspaceTaskLifecycleState | string
  recovery_actions?: string[]
  download_url?: string
  data?: Record<string, any>
  raw?: any
}

export const workspaceSimpleChatStreamAPI = async (
  data: WorkSpaceSimpleTask,
  handlers: {
    onMessage?: (chunk: string) => void
    onFinalMessage?: (message: string) => void
    onEvent?: (event: WorkspaceStreamEvent) => void
    onError?: (err: any) => void
    onClose?: () => void
  }
) => {
  const token = localStorage.getItem('token')
  const ctrl = new AbortController()
  const streamUrl = apiUrl('/api/v1/workspace/simple/chat')

  try {
    await fetchEventSource(streamUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: token ? `Bearer ${token}` : '',
      },
      body: JSON.stringify(data),
      signal: ctrl.signal,
      openWhenHidden: true,
      onmessage(event) {
        if (!event.data) return

        try {
          const parsed = JSON.parse(event.data)
          if (parsed?.event !== undefined) {
            const normalized: WorkspaceStreamEvent = {
              id: crypto.randomUUID(),
              type: String(parsed.event || 'event'),
              title: String(
                parsed?.data?.message ||
                  parsed?.data?.summary ||
                  parsed?.data?.phase ||
                  parsed.event ||
                  'event'
              ),
              detail: String(
                parsed?.data?.result ||
                  parsed?.data?.error ||
                  parsed?.data?.chunk ||
                  parsed?.data?.accumulated ||
                  ''
              ),
              isFinal: parsed?.event === 'final',
              task_id: parsed?.data?.task_id,
              trace_id: parsed?.data?.trace_id,
              artifact_id: parsed?.data?.artifact_id,
              citation_ids: parsed?.data?.citation_ids,
              tool_id: parsed?.data?.tool_id,
              tool_call_id: parsed?.data?.tool_call_id,
              approval_id: parsed?.data?.approval_id,
              required_approval: parsed?.data?.required_approval,
              audit_ref: parsed?.data?.audit_ref,
              lifecycle_state: parsed?.data?.lifecycle_state,
              recovery_actions: parsed?.data?.recovery_actions,
              download_url: parsed?.data?.download_url,
              data: parsed?.data || {},
              raw: parsed,
            }
            handlers.onEvent?.(normalized)
          }

          if (parsed?.event === 'final') {
            if (parsed?.data?.done === true && typeof parsed?.data?.accumulated === 'string' && parsed.data.accumulated !== '') {
              handlers.onFinalMessage?.(parsed.data.accumulated)
              return
            }
            if (parsed?.data?.chunk !== undefined && parsed.data.chunk !== '') {
              handlers.onMessage?.(parsed.data.chunk)
              return
            }
            if (parsed?.data?.message !== undefined && parsed.data.message !== '') {
              handlers.onMessage?.(parsed.data.message)
              return
            }
          } else if (typeof parsed?.data?.accumulated === 'string' && parsed.data.accumulated !== '') {
            return
          } else if (typeof parsed?.data?.message === 'string' && parsed.data.message !== '') {
            return
          }
        } catch {
          const rawText = event.data.trim()
          if (rawText && !rawText.startsWith('{') && !rawText.startsWith('[')) {
            handlers.onMessage?.(rawText)
          }
        }
      },
      onerror(err) {
        handlers.onError?.(err)
        ctrl.abort()
      },
      onclose() {
        handlers.onClose?.()
      },
    })
  } catch (error: any) {
    if (error?.name !== 'AbortError') {
      handlers.onError?.(error)
    }
  }
}

export const createWorkspaceTaskAPI = async (data: WorkSpaceSimpleTask) => {
  return request({
    url: '/api/v1/workspace/task',
    method: 'post',
    data,
  })
}

export const getWorkspaceTaskLifecycleAPI = async () => {
  return request({
    url: '/api/v1/workspace/task-lifecycle',
    method: 'get',
  })
}

export const createWorkspaceFileAPI = async (data: WorkspaceFileCreateRequest) => {
  return request({
    url: '/api/v1/workspace/file',
    method: 'post',
    data,
  })
}

export const createWorkspaceIngestAPI = async (data: WorkspaceIngestRequest) => {
  return request({
    url: '/api/v1/workspace/ingest',
    method: 'post',
    data,
  })
}

export const getWorkspaceTaskAPI = async (taskId: string) => {
  return request({
    url: `/api/v1/workspace/task/${taskId}`,
    method: 'get',
  })
}

export const getWorkspaceTaskEventsAPI = async (taskId: string) => {
  return request({
    url: `/api/v1/workspace/task/${taskId}/events`,
    method: 'get',
  })
}

export const workspaceTaskEventsStreamAPI = async (
  taskId: string,
  handlers: {
    onEvent?: (event: WorkspaceStreamEvent) => void
    onError?: (err: any) => void
    onClose?: () => void
  }
) => {
  const token = localStorage.getItem('token')
  const ctrl = new AbortController()
  const streamUrl = apiUrl(`/api/v1/workspace/task/${taskId}/events/stream`)

  try {
    await fetchEventSource(streamUrl, {
      method: 'GET',
      headers: {
        Authorization: token ? `Bearer ${token}` : '',
      },
      signal: ctrl.signal,
      openWhenHidden: true,
      onmessage(event) {
        if (!event.data) return
        try {
          const parsed = JSON.parse(event.data)
          const normalized: WorkspaceStreamEvent = {
            id: String(parsed?.data?.event_id || crypto.randomUUID()),
            type: String(parsed?.event || parsed?.data?.phase || 'event'),
            title: String(parsed?.data?.message || parsed?.event || 'event'),
            detail: String(parsed?.data?.result || parsed?.data?.error || parsed?.data?.chunk || ''),
            isFinal: parsed?.event === 'task_completed' || parsed?.data?.status === 'completed',
            task_id: parsed?.data?.task_id,
            trace_id: parsed?.data?.trace_id,
            artifact_id: parsed?.data?.artifact_id,
            citation_ids: parsed?.data?.citation_ids,
            tool_id: parsed?.data?.tool_id,
            tool_call_id: parsed?.data?.tool_call_id,
            approval_id: parsed?.data?.approval_id,
            required_approval: parsed?.data?.required_approval,
            audit_ref: parsed?.data?.audit_ref,
            lifecycle_state: parsed?.data?.lifecycle_state,
            recovery_actions: parsed?.data?.recovery_actions,
            download_url: parsed?.data?.download_url,
            data: parsed?.data || {},
            raw: parsed,
          }
          handlers.onEvent?.(normalized)
        } catch (err) {
          handlers.onError?.(err)
        }
      },
      onerror(err) {
        handlers.onError?.(err)
        ctrl.abort()
      },
      onclose() {
        handlers.onClose?.()
      },
    })
  } catch (error: any) {
    if (error?.name !== 'AbortError') {
      handlers.onError?.(error)
    }
  }
}

export const approveWorkspaceTaskAPI = async (taskId: string, data: WorkspaceApprovalRequest) => {
  return request({
    url: `/api/v1/workspace/task/${taskId}/approve`,
    method: 'post',
    data,
  })
}

export const cancelWorkspaceTaskAPI = async (taskId: string, data: WorkspaceCancelRequest) => {
  return request({
    url: `/api/v1/workspace/task/${taskId}/cancel`,
    method: 'post',
    data,
  })
}

export const getWorkspaceArtifactAPI = async (artifactId: string) => {
  return request({
    url: `/api/v1/workspace/artifact/${artifactId}`,
    method: 'get',
  })
}

export const downloadWorkspaceArtifactAPI = async (artifactId: string) => {
  return request({
    url: `/api/v1/workspace/artifact/${artifactId}/download`,
    method: 'get',
    responseType: 'blob',
  })
}

export const createWorkspaceFeedbackAPI = async (data: WorkspaceFeedbackRequest) => {
  return request({
    url: '/api/v1/workspace/feedback',
    method: 'post',
    data,
  })
}
